torch/csrc/jit/serialization/storage_context.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+namespace jit {
+
+// Used in torch.package and TorchScript serialization to coordinate
+// sharing of storages between models. Also used to create deterministic
+// naming for storages.
+class TORCH_API SerializationStorageContext {
+ public:
+  explicit SerializationStorageContext() : unique_id(0) {}
+
+  uint64_t addStorage(c10::Storage storage) {
+    TORCH_INTERNAL_ASSERT(!hasStorage(storage));
+    uint64_t id = unique_id++;
+    storage_id_map_.insert({storage, id});
+    return id;
+  }
+
+  bool hasStorage(c10::Storage storage) {
+    return storage_id_map_.find(storage) != storage_id_map_.end();
+  }
+
+  uint64_t getId(c10::Storage storage) {
+    TORCH_INTERNAL_ASSERT(hasStorage(storage));
+    return storage_id_map_.find(storage)->second;
+  }
+
+  ~SerializationStorageContext() = default;
+
+ private:
+  class StorageSerializationComparator {
+   public:
+    bool operator()(const c10::Storage& lhs, const c10::Storage& rhs) const {
+      return rhs.unsafeGetStorageImpl() < lhs.unsafeGetStorageImpl();
+    }
+  };
+
+  uint64_t unique_id;
+  std::map<c10::Storage, uint64_t, StorageSerializationComparator>
```

Comment:
small comments:
1. `std::unordered_map` is slightly better here, since it has better lookup/retrieval performance and we don't care about ordering.
2. We don't need `unique_id`. Since we never delete storage from the serialization context, we can just do `storage_id_map_[storage] = storage_id_map_.size()`.